Design doc for SSO (#3425)
scality/metalk8s: docs/developer/architecture/sso.rst

Single Sign On
==============

Context
-------

> As for now, when we log in into the PlatformUI then we click on grafana,

Reviewer: Not sure about it, but In this context, I think we are talking about MetalK8s UI.
And we don't click on Grafana, we click **advanced metrics** or we access to Grafana.
To me, it's `Log in to` or `log into`, but not both.
```suggestion
As for now, when we log in to the MetalK8s UI and then we access to Grafana,
```

we have to log in again because these two components do not share a common
session and Dex (the authentication provider) does not offer such

> fonctionality.

Reviewer: ```suggestion
functionality.
```

Requirements
------------
* Ability to share login session between UI components
* Ability to logout once and for all UI components

User Stories
------------

> As an application administrator I want to:

Reviewer: ```suggestion
As a Platform Administrator, I want to:
```

- While logged in, if I click on "advanced metrics" I want to access Grafana
  without logging in again
- While logged in, If I click logout I want to be logged out from all other

>   components

Reviewer: ```suggestion
- While logged in, if I click **advanced metrics**, I want to access Grafana
  without logging in again
- While logged in, If I click **logout**, I want to be logged out from all the
  components
```

- This functionality should be available in both static password and other

>   authentication providers

Reviewer: It's not part of a user story, it's more a requirement.

Design Choices
--------------

Rejected Designs
****************

Custom Keycloak

> ~~~~~~~~~~~~~~~

Reviewer: Why custom ?
```suggestion
Keycloak
~~~~~~~~
```
Or are we talking about using Keycloak instead of Dex ? It's not clear here or in the following paragraph.

> Keycloak does persist a session and thus can support an SSO behavior, But using

Reviewer: ```suggestion
Keycloak does persist a session and thus can support an SSO behavior, but using
```

it means maintaining a database while we want a simple authentication provider
to fallback on.

Deprecate Dex

> ~~~~~~~~~~~~~

Reviewer: Should be renamed to something like "Replace Dex by oauth2-proxy" as the rejected choice is not the Dex deprecation, but more the usage of oauth2-proxy alone because of its limitations.

The issue of SSO is that Dex does not keep track of the session, we may as well
get rid of Dex and replace it with other mechanism.
The issue with this solution is that the replacement, oauth2-proxy, does not

> provide oidc functionalities and thus cannot be used by the kube-apiserver.

Reviewer 2: Maybe we should detail why authenticating using a proxy on kube-apiserver (https://kubernetes.io/docs/reference/access-authn-authz/authentication/#authenticating-proxy) is not an option we retained

> provide oidc functionalities and thus cannot be used by the kube-apiserver.

Reviewer: It should be reworded:
```suggestion
Although oauth2-proxy keeps track of sessions unlike Dex, it does not provide OIDC
functionnalities and thus cannot be used by the kube-apiserver.
```
The first sentence also works for the section above, but it's kind of redundant with what's explain in the "Context" section anyway.

Chosen Approach
***************

We will use a combination of Dex and Oauth2-proxy,

- Dex is used to provide token from static password authentication and to proxy
  other configured OIDC providers
- Oauth2-proxy is used to handle authentication and session support for the
  different endpoints

Implementations details
-----------------------

Components functionalities
**************************

Dex will:

- Provides static password authentication support

> - Provides tokens backed by the static password, kube-apiserver will be

Reviewer: ```suggestion
- Provide static password authentication support
- Provide tokens backed by the static password, kube-apiserver will be
```

  configured to talk to Dex

> - Proxy other OIDC providers like Keycloak

Reviewer: Dex does not only support OIDC, but also LDAP, AuthProxy, SAML, etc.

Oauth2-proxy will:

> - Be installed as a side car authentication layer for the control plane

Reviewer: ```suggestion
- Be installed as a sidecar authentication layer for the control plane
```

  components (PlatformUI, Grafana, Prometheus, etc)

> - Provide session support, moving from one components to another while keeping

Reviewer: ```suggestion
- Provide session support, moving from one component to another while keeping
```

  the same session id
- Redirect to Dex and recover the authentication token if the session does not

>   exist or no more valid

Reviewer: ```suggestion
  exist or is not valid anymore (e.g. expired)
```

> - Provides the token to the protected upstream via headers or in the session

Reviewer: ```suggestion
- Provide the token to the protected upstream via headers or in the session
```

  itself (some logic needed to extract the token from the session)

Authentication workflow
***********************

> 1. User visits endpoint-1 (exp PlatformUI)

Reviewer: ```suggestion
1. User visits endpoint-1 (e.g. PlatformUI)
```

2. Oauth2-proxy finds no session cookies, so it redirect the User to Dex
   in order to authenticate

> 3. User authenticate (with static password or through another OIDC provider)

Reviewer: ```suggestion
3. User authenticates (with static password or through another OIDC provider)
```

4. Oauth2-proxy gets a token and creates a session
5. Oauth2-proxy passes the token to the upstream, User is logged in

> 6. User visits endpoint-2 (exp Grafana)

Reviewer: ```suggestion
6. User visits endpoint-2 (e.g. Grafana)
```

7. Oauth2-proxy reads the session, and redirect the User to the endpoint-2
   while passing the token to the upstream
8. User is automatically logged in the endpoint-2

> 9. User log out, Oauth2-proxy deletes the session

Reviewer: ```suggestion
9. User logs out, Oauth2-proxy deletes the session
```

10. GOTO (1)

NB: Ajax requests will receive 401 if the request does not contain the cookie.

Upstream authentication
***********************

Using Oauth2-proxy, upstreams will receive the authentication token as headers.
In case of client-side application, it should send a request to ``/userinfo``

> using the cookie and then retreive the headers from the response.

Reviewer: ```suggestion
using the cookie and then retrieve the headers from the response.
```

The headers contain:

- The user claim (X-Forwarded-User)
- The user groups if configured (X-Forwarded-Groups)
- The email claim (X-Forwarded-Email)
- The full OIDC access token (X-Forwarded-Access-Token)

Operating Oauth2-proxy
**********************

Oauth2-proxy serves some built-in endpoints, from which

- ``/ping`` returns 200 OK if it is working correctly which can be used in
  container healthchecks

> - ``/metrics`` provides metrics in prometheus format.

Reviewer 2: We might want to add a `Monitoring Oauth2-proxy` section to define which prometheus rules we would want to deploy to be alerted in case of unavailability of Oauth2-Proxy

   while passing the token to the upstream
8. User is automatically logged in the endpoint-2
9. User log out, Oauth2-proxy deletes the session
10. GOTO (1)

NB: Ajax requests will receive 401 if the request does not contain the cookie.

Upstream authentication
***********************

Using Oauth2-proxy, upstreams will receive the authentication token as headers.
In case of client-side application, it should send a request to ``/userinfo``
using the cookie and then retreive the headers from the response.

The headers contain:

- The user claim (X-Forwarded-User)
- The user groups if configured (X-Forwarded-Groups)
- The email claim (X-Forwarded-Email)
- The full OIDC access token (X-Forwarded-Access-Token)

Operating Oauth2-proxy
**********************

Oauth2-proxy serves some built-in endpoints, from which

- ``/ping`` returns 200 OK if it is working correctly which can be used in
  container healthchecks
- ``/metrics`` provides metrics in prometheus format.
- ``/userinfo`` provides user and email claims
- ``/oauth2/sign_out`` removes the session to sign out the user

Monitoring Oauth2-proxy
***********************

> Oauth2-proxy exposes the follwoing metrics:

Reviewer: ```suggestion
Oauth2-proxy exposes the following metrics:
```

- oauth2_proxy_requests_total: Total number of requests by HTTP status code
  (Counter)
- oauth2_proxy_response_duration_seconds: A histogram of request latencies
  (Histogram)
- oauth2_proxy_requests_in_flight: Current number of requests being served

>   (Gauge)

Reviewer: Ok, great, but what are we gonna do with these metrics ? Create a Grafana dashboard ? Create some Prometheus rules ?

Alerting should be based on the status of the Pod (when readiness/liveness)

> probes are correctly set up based on the ``/ping`` endpoing.

Reviewer: ```suggestion
probes are correctly set up based on the ``/ping`` endpoint.
```
